neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/embedded/impl/EmbeddedNeo4jGraphDatabaseFactory.java

```diff
@@ -86,4 +142,83 @@ private void setConfigurationFromLocation(GraphDatabaseBuilder builder, URL cfgL
 			builder.loadPropertiesFromURL( cfgLocation );
 		}
 	}
+
+	static void shutdownGraphDatabaseService(GraphDatabaseService neo4jDb) {
+		GraphDatabaseAPI sq = (GraphDatabaseAPI) neo4jDb;
+		String key = sq.getStoreDir().getAbsolutePath();
+		if ( GRAPH_DATABASE_SERVICE_MAP.containsKey( key ) ) {
+			synchronized (GRAPH_DATABASE_SERVICE_MAP) {
+				FactoryHolder factoryHolder = GRAPH_DATABASE_SERVICE_MAP.get( key );
+				factoryHolder.setCounter( factoryHolder.getCounter() - 1 );
+				LOG.debugf( " Counter : %s  for path %s", factoryHolder.getCounter(), key );
+
+				GraphDatabaseFacade graphDatabaseFacade = factoryHolder.getGraphDatabaseFacade();
+				try {
+					GraphDatabaseFacade.SPI spi = null;
+					Object obj = getPrivateField( GraphDatabaseFacade.class.getDeclaredField( "spi" ), graphDatabaseFacade );
+					spi = (GraphDatabaseFacade.SPI) obj;
+					LOG.debugf( " isInOpenTransaction: %s", spi.isInOpenTransaction() );
+					if ( spi.isInOpenTransaction() ) {
+						LOG.warnf( "currentTransaction  %s started at %s", spi.currentTransaction() , spi.currentTransaction().startTime() );
+						spi.currentTransaction().failure();
+						spi.currentTransaction().closeTransaction();
+						throw LOG.existsOpenTransactionsFornDatabasePathException( key );
+					}
+				}
+				catch ( IllegalAccessException | NoSuchFieldException e) {
+					throw new HibernateException( "Can't get value of private variable by reflection!",e );
+				}
+				catch (TransactionFailureException e) {
+					throw new HibernateException( "Can't close transction!",e );
+				}
+
+
+				if ( factoryHolder.getCounter() == 0 ) {
+					LOG.infof( " Shutdown db for path : %s", key );
+					factoryHolder.getGraphDatabaseFacade().shutdown();
+					GRAPH_DATABASE_SERVICE_MAP.remove( key );
+				}
+			}
+		}
+		else {
+			throw LOG.unknownDatabasePathException( key );
+		}
+	}
+
+	private static Object getPrivateField(Field privateField, Object targetObject) throws IllegalAccessException {
+		privateField.setAccessible( true );
+		return privateField.get( targetObject );
+	}
+
+
+	private static class FactoryHolder {
+		private int counter;
+		private GraphDatabaseFacade graphDatabaseFacade;
+		private boolean newHolder;
```

Comment:
are you sure you need this field?

Reply:
I need mark new instance.
If the instance is not new I need increment counter of clients.

Reply:
Maybe you can use the counter, it's an idea. WDYT?